### PR TITLE
[router][grpc] Fix tool call streaming bugs: empty tool names, state pollution, and panics

### PR DESCRIPTION
**Pull request by CatherineSue**
## Motivation

This PR fixes critical bugs in the Rust tool parser streaming implementation to achieve parity with the Python implementation, particularly around:

1. **Empty tool name bug**: Parser incorrectly returned empty tool names when receiving incomplete JSON like `{"name": "`
2. **State pollution in streaming**: Pooled parsers caused state leakage between requests
3. **Argument diff calculation**: Incorrect logic for calculating incremental argument changes
4. **Parser creation panics**: Router panicked when no parser was found for a model
5. **Excessive warning logs**: Warning messages repeated for every chunk instead of once per request

These bugs caused incorrect tool call parsing in streaming scenarios, especially with realistic token-by-token LLM output.

---

### 1. Fix Empty Tool Name Bug with `allow_partial_strings` Parameter

**Problem**: When streaming `{"name": "search"...}` char-by-char, the parser received `{"name": "` and incorrectly parsed it as a tool call with empty name `""`.

**Root Cause**: Rust's `partial_json.rs` didn't support Python's flag mechanism (`Allow.ALL & ~Allow.STR`) that controls whether incomplete strings should be parsed. Python disables partial string parsing when `current_tool_name_sent=False` to prevent this exact bug.

**Fix** (commit 839389855):
- Added `allow_partial_strings: bool` parameter to `PartialJson::parse_value()`
- When `false`, incomplete strings like `"` or `"abc` are not treated as valid values
- Updated all parsers to pass `allow_partial_strings=false` when parsing tool names before they're sent

---

### 2. Per-Request Parser Approach (Eliminate State Pollution)

**Problem**: Parsers were pooled and reused across requests, causing state pollution. Parser internal state (buffers, flags, counters) from request A leaked into request B.

**Fix** (commits fed66f846, fb064eb3a):

**Added `reset()` method to all parsers**:
```rust
pub trait ToolParser: Send + Sync {
    fn reset(&mut self);  // Clear all internal state
    // ... other methods
}
```

**Changed factory pattern**:
- **Before**: `get_parser(model)` → returned pooled parser (shared state)
- **After**: `create_parser(model)` → creates fresh instance per request (isolated state)
---

### 3. Naming Convention Changes (Idiomatic Rust)

**Problem**: Names like `ToolParserFactory` and `ReasoningParserFactory` included redundant parent module prefix, violating Rust conventions.

**Fix** (commit 84305abb3):
- `ToolParserFactory` → `ParserFactory` (in `tool_parser` module)
- `ReasoningParserFactory` → `ParserFactory` (in `reasoning_parser` module)
- `ToolParserError` → `ParserError`
- `parse_streaming_incremental` → `parse_incremental` (removed redundant "streaming")
- Similarly for all related types

**Rust Convention**: Module-scoped types don't need parent module in name:
```rust
// BEFORE (verbose)
use tool_parser::ToolParserFactory;
let factory: ToolParserFactory = ...;

// AFTER (idiomatic)
use tool_parser::ParserFactory;
let factory: ParserFactory = ...;
```

---

### 4. Fix `prev_tool_call_arr` Update Logic

**Problem**: `prev_tool_call_arr` (tracks previous tool call JSON for diff calculation) was updated incorrectly, causing argument diffs to be wrong or duplicated.

**Fix** (commit 4f5580b29):
- Update `prev_tool_call_arr` **only after** successfully emitting a new delta
- Ensure it reflects the exact JSON state that was last sent to client
- Properly handle completion state to avoid redundant updates

---

### 5. Fix Argument Diff Calculation

**Problem**: The `args_diff` had multiple issues:
1. Didn't properly compute JSON diff between current and previous arguments
2. Caused duplicate or missing argument chunks in streaming

**Fix** (commit 0aa33a451):
- Add `prefix` to make sure we only send diff that has been verified

**Files Changed**:
- `sgl-router/src/tool_parser/parsers/helpers.rs` (+64 lines)

---

### 6. Fix `streamed_args_for_tool` Management

**Problem**: `streamed_args_for_tool` (tracks which tools had args streamed) was cleared prematurely, causing loss of state.

**Fix** (commit eb13f4ecd):
- Only update `streamed_args_for_tool` when tool call is complete
- Preserve state across incremental parsing calls
- Properly synchronize with `current_tool_id`

---

### 7. Fix Parser Not Found Panic + Optimize Parser Availability Check

**Fix** (commits 2f81ad830 + latest):

- Added efficient availability check methods: `has_parser`, `has_parser_for_model`
- Check availability once at request start
- Skip parser processing entirely if unavailable**:

**Files Changed**:
- `sgl-router/src/routers/grpc/streaming.rs`
- `sgl-router/src/tool_parser/factory.rs`
- `sgl-router/src/reasoning_parser/factory.rs`

---

### 8. Update Streaming Tests

**Added** (commits 6bac5c46d, f5e9eef95):
- Realistic streaming tests that simulate actual LLM token-by-token output (2-5 char chunks)
- Critical test case for empty tool name bug: `test_json_bug_incomplete_tool_name_string()`
- Tests for all parsers (JSON, Llama, Qwen) with strategic chunk boundaries
- Edge cases: long URLs, Unicode, XML/Python tags split across chunks

---

## Accuracy Tests

n = 1, streaming = True, llama-3.1-8b-instruct

Screenshot 2025-10-09 at 12 02 45 AM

n = 2

Screenshot 2025-10-09 at 12 06 47 AM

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: sgl-router/src/tool_parser/parsers/helpers.rs [hunk 1]
```diff
@@ -14,6 +14,27 @@ pub fn get_tool_indices(tools: &[Tool]) -> HashMap<String, usize> {
...
+    for i in 0..min_len {
+        if chars1[i] == chars2[i] {
+            prefix.push(chars1[i]);
+        } else {
+            break;
+        }
+    }
+
+    prefix
+}
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply sgl-router/src/tool_parser/parsers/helpers.rs:25]: Good suggestion, applied

**Comment by CatherineSue** [comment]
Manual test:

Screenshot 2025-10-09 at 12 35 04 AM

**Review by slin1237 — APPROVED** [review]
